bundles/org.eclipse.ui.ide/extensions/org/eclipse/ui/actions/RefreshAction.java

```diff
@@ -113,11 +115,14 @@ void checkLocationDeleted(IProject project) throws CoreException {
 		if (!project.exists()) {
 			return;
 		}
-		IFileInfo location = IDEResourceInfoUtils.getFileInfo(project.getLocationURI());
+		URI locationURI = project.getLocationURI();
+		IFileInfo location = IDEResourceInfoUtils.getFileInfo(locationURI);
 		if (!location.exists()) {
+			IFileStore fileStore = IDEResourceInfoUtils.getFileStore(locationURI);
+			String projectPath = fileStore != null ? fileStore.toString() : location.getName();
 			String message = NLS.bind(
 					IDEWorkbenchMessages.RefreshAction_locationDeletedMessage,
-					project.getName(), location.toString());
+					project.getName(), projectPath);
```

Comment:
I don't see a "quick" way to separate this UI code from the logic, but at least it would make sense to extract this into a method, like this:

```java
if (!location.exists()) {
    String displayedProjectPath = generatePath(project, location, filestore);
    String message = NLS.bind(
					IDEWorkbenchMessages.RefreshAction_locationDeletedMessage,
					project.getName(), projectPath);
```

What do you think?

Reply:
Sure, will change.